drivers/clock_control/clock_fixed_factor.c

```diff
@@ -0,0 +1,104 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (C) 2022 Google, LLC
+ *
+ */
+
+#include <zephyr/drivers/clock_control.h>
+
+#define DT_DRV_COMPAT fixed_factor_clock
+
+struct fixed_factor_clock_config {
+	const struct device *clk_dev;
+	clock_control_subsys_t clk_subsys;
+	uint32_t divider;
+	uint32_t multiplier;
+};
+
+static int fixed_factor_clk_init(const struct device *dev)
+{
+	const struct fixed_factor_clock_config *config = dev->config;
+
+	return clock_control_on(config->clk_dev, config->clk_subsys);
+}
+
+static int fixed_factor_clk_on(const struct device *dev,
+			       clock_control_subsys_t sys)
+{
+	const struct fixed_factor_clock_config *config = dev->config;
+
+	ARG_UNUSED(sys);
+
+	return clock_control_on(config->clk_dev, config->clk_subsys);
+}
+
+static int fixed_factor_clk_off(const struct device *dev,
+				clock_control_subsys_t sys)
+{
+	const struct fixed_factor_clock_config *config = dev->config;
+
+	ARG_UNUSED(sys);
+
+	return clock_control_off(config->clk_dev, config->clk_subsys);
+}
+
+static enum clock_control_status fixed_factor_clk_get_status(const struct device *dev,
+							     clock_control_subsys_t sys)
+{
+	const struct fixed_factor_clock_config *config = dev->config;
+
+	return clock_control_get_status(config->clk_dev, config->clk_subsys);
+}
+
+static int fixed_factor_clk_get_rate(const struct device *dev,
+				     clock_control_subsys_t sys,
+				     uint32_t *rate)
+{
+	const struct fixed_factor_clock_config *config = dev->config;
+	uint32_t input_rate;
+	int err;
+
+	ARG_UNUSED(sys);
+
+	err = clock_control_get_rate(config->clk_dev, config->clk_subsys,
+				     &input_rate);
+	if (err) {
+		return err;
+	}
+
+	*rate = ((uint64_t)input_rate * config->multiplier) / config->divider;
+
+	return 0;
+}
+
+static const struct clock_control_driver_api fixed_factor_clk_api = {
+	.on = fixed_factor_clk_on,
+	.off = fixed_factor_clk_off,
+	.get_status = fixed_factor_clk_get_status,
+	.get_rate = fixed_factor_clk_get_rate
+};
+
+#define INPUT_SUBSYS_OR_ALL(idx) \
+	COND_CODE_1(DT_INST_NUM_CLOCKS(idx), \
+		(DT_INST_PHA_BY_IDX_OR(idx, clocks, 0, name, \
+				      CLOCK_CONTROL_SUBSYS_ALL)), \
+		    (CLOCK_CONTROL_SUBSYS_ALL))
+
+#define FIXED_FACTOR_CLK_INIT(idx) \
+	static const struct fixed_factor_clock_config fixed_factor_clock_config_##idx = { \
+		.clk_dev = DEVICE_DT_GET(DT_INST_PHANDLE(idx, clocks)), \
+		.clk_subsys = (clock_control_subsys_t)INPUT_SUBSYS_OR_ALL(idx), \
+		.multiplier = DT_INST_PROP(idx, clock_mult), \
+		.divider = DT_INST_PROP(idx, clock_div), \
+	}; \
+	DEVICE_DT_INST_DEFINE(idx, \
+		fixed_factor_clk_init, \
+		NULL, \
+		NULL, \
+		&fixed_factor_clock_config_##idx, \
+		PRE_KERNEL_1, \
+		CONFIG_CLOCK_CONTROL_INIT_PRIORITY, \
```

Comment:
About the priority:
Since client of `clock_control` API, it should be done after init of clock_control driver.
But yet, at least on STM32, we need fixed clocks to be initialized in the early phase since some other clocks will depend on fixed clocks availability (PLL for instance).
So we'd need similar driver enabled for PLL as well, that would be initialized after fixed (factor) clocks.
Then a driver for sysclock which usually requires PLL.
And finally you'll need to configure peripheral busses.
(Note that this is a very simple approach as you usually also need to perform other actions like update Flash latency before and/or after this whole sequence)

Taking this into account, I wonder about the whole plan for getting this driver to be really useful.
Or we should assume that it won't ever be compatible with some platforms.

@mfischer can you comment on that part?

Note: I'm setting this comment as "blocking" to trigger the discussion, but it is not meant to be blocking in the end or even require changes (depending on the output of the discussion).

Reply:
@erwango I've updated change set to simplify the fixed-rate clock driver, I don't think it should have any external dependencies left. What would be the earliest priority you'd want the driver to initialize?

Note: For a fixed-rate clock you could in theory always bypass the clock control API and pull the rate / frequency out of DT, since, by definition it's always on and has a fixed rate.

The use-case I'm trying to enable is to allow for drivers to runtime derive their input clock frequency - independently if they're being fed by a fixed-rate clock or another clock control API provider.

Reply:
> @erwango I've updated change set to simplify the fixed-rate clock driver, I don't think it should have any external dependencies left. What would be the earliest priority you'd want the driver to initialize?

Assuming: 
- You don't need clock_control API at driver init
- Fixed clocks are needed for core clock configuration (as part of the core clock tree)
This driver needs to be initialized before the piece of code configuring core clock (`clock_control` in STM32 case).

> The use-case I'm trying to enable is to allow for drivers to runtime derive their input clock frequency - independently if they're being fed by a fixed-rate clock or another clock control API provider.

This is already possible with existing API. What are the limitation you're seeing ?

Reply:
> This is already possible with existing API. What are the limitation you're seeing ?

You cannot call `clock_control_get_rate()` on a fixed (factor) clock in the tree as it is today, thus requiring all drivers with a `clocks` devicetree property having to check for a fixed (factor) clock at compile-time and use the clock rate directly - or defer to run-time and call `clock_control_get_rate()`.

Reply:
Ok, right. I'm still assuming that we only have one `clock_control` driver centralizing all clock related function (which makes this possible), but I understand this will tend to be old fashioned in medium term.

Reply:
@erwango if you want to keep the same binding for fixed clocks I'm not sure how you'd do that with a single driver. Can you elaborate?

Reply:
Well this is how it it handled today in STM32 clock_control driver for instance. There are multiple clocks compatibles, but only one driver handles the whole information to build the clock tree.

I understand that this is a shift of paradigm and for this driver to be usable in most clocks tree it requires a complete clock framework around it.
Let me synthesize the assumptions around this driver:
- Clock initialization should be done prior to the use of this driver
- It cannot be called in the core clock tree initialization phase if such phase is done as part of a 'main' clock_control driver init.
Is that correct ?